File: src/main/java/org/xodium/vanillaplus/modules/DoorsModule.java
package org.xodium.vanillaplus.modules;

import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;
import java.util.Optional;

import org.bukkit.event.Event;
import org.bukkit.Bukkit;
import org.bukkit.GameMode;
import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.NamespacedKey;
import org.bukkit.Registry;
import org.bukkit.Sound;
import org.bukkit.SoundCategory;
import org.bukkit.World;
import org.bukkit.block.Block;
import org.bukkit.block.BlockFace;
import org.bukkit.block.data.Bisected;
import org.bukkit.block.data.BlockData;
import org.bukkit.block.data.Openable;
import org.bukkit.block.data.type.Door;
import org.bukkit.block.data.type.Gate;
import org.bukkit.block.data.type.TrapDoor;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.block.Action;
import org.bukkit.event.player.PlayerInteractEvent;
import org.bukkit.inventory.EquipmentSlot;
import org.bukkit.scheduler.BukkitRunnable;
import org.xodium.vanillaplus.Database;
import org.xodium.vanillaplus.VanillaPlus;
import org.xodium.vanillaplus.interfaces.Modular;
import org.xodium.vanillaplus.interfaces.PERMS;
import org.xodium.vanillaplus.records.AdjacentBlockRecord;

import com.google.common.base.Enums;

// TODO: refactor.
public class DoorsModule implements Modular {
    private final String className = getClass().getSimpleName();
    private final VanillaPlus vp = VanillaPlus.getInstance();
    private final Database db = new Database();

    // Sound settings
    public static final String SOUND_KNOCK_CATEGORY = ".sound_knock_category";
    public static final String SOUND_KNOCK_PITCH = ".sound_knock_pitch";
    public static final String SOUND_KNOCK_VOLUME = ".sound_knock_volume";
    public static final String SOUND_KNOCK_WOOD = ".sound_knock_wood";

    // Behavior settings
    public static final String ALLOW_AUTOCLOSE = ".allow_autoclose";
    public static final String ALLOW_DOUBLEDOORS = ".allow_doubledoors";
    public static final String ALLOW_KNOCKING = ".allow_knocking";
    public static final String ALLOW_KNOCKING_GATES = ".allow_knocking_gates";
    public static final String ALLOW_KNOCKING_TRAPDOORS = ".allow_knocking_trapdoors";
    public static final String KNOCKING_REQUIRES_EMPTY_HAND = ".knocking_requires_empty_hand";
    public static final String KNOCKING_REQUIRES_SHIFT = ".knocking_requires_shift";

    // Auto-close settings
    public static final String AUTOCLOSE_DELAY = ".autoclose_delay";

    // Others
    private final HashMap<Block, Long> autoClose = new HashMap<>();
    private final static AdjacentBlockRecord[] POSSIBLE_NEIGHBOURS = {
            new AdjacentBlockRecord(0, -1, Door.Hinge.RIGHT, BlockFace.EAST),
            new AdjacentBlockRecord(0, 1, Door.Hinge.LEFT, BlockFace.EAST),

            new AdjacentBlockRecord(1, 0, Door.Hinge.RIGHT, BlockFace.SOUTH),
            new AdjacentBlockRecord(-1, 0, Door.Hinge.LEFT, BlockFace.SOUTH),

            new AdjacentBlockRecord(0, 1, Door.Hinge.RIGHT, BlockFace.WEST),
            new AdjacentBlockRecord(0, -1, Door.Hinge.LEFT, BlockFace.WEST),

            new AdjacentBlockRecord(-1, 0, Door.Hinge.RIGHT, BlockFace.NORTH),
            new AdjacentBlockRecord(1, 0, Door.Hinge.LEFT, BlockFace.NORTH)
    };

    {
        Bukkit.getScheduler().runTaskTimer(vp, () -> {
            Iterator<Map.Entry<Block, Long>> it = autoClose.entrySet().iterator();
            while (it.hasNext()) {
                Map.Entry<Block, Long> entry = it.next();
                Block b = entry.getKey();
                Long time = entry.getValue();
                if (System.currentTimeMillis() < time)
                    continue;
                if (b.getBlockData() instanceof Openable) {
                    Openable openable = (Openable) b.getBlockData();
                    if (openable.isOpen()) {
                        if (openable instanceof Door) {
                            Block otherDoor = this.getOtherPart((Door) openable, b);
                            if (otherDoor != null) {
                                this.toggleOtherDoor(b, otherDoor, false);
                            }
                        } else if (openable instanceof Gate) {
                            b.getWorld().playSound(b.getLocation(), Sound.BLOCK_FENCE_GATE_CLOSE, 1.0f, 1.0f);
                        }
                        openable.setOpen(false);
                        b.setBlockData(openable);
                        b.getWorld().playSound(b.getLocation(), Sound.BLOCK_IRON_DOOR_CLOSE, 1.0f, 1.0f);
                    }
                }
                it.remove();
            }
        }, 1, 1);
    }

    @EventHandler(priority = EventPriority.MONITOR)
    public void onRightClick(PlayerInteractEvent e) {
        Block clickedBlock = e.getClickedBlock();
        if (clickedBlock == null)
            return;
        BlockData blockData = clickedBlock.getBlockData();

        if (e.getHand() != EquipmentSlot.HAND
                || e.getAction() != Action.RIGHT_CLICK_BLOCK
                || e.useInteractedBlock() == Event.Result.DENY
                || e.useItemInHand() == Event.Result.DENY
                || !e.getPlayer().hasPermission(PERMS.DOORSMODULE.USE)
                || !(blockData instanceof Door || blockData instanceof Gate))
            return;

        if (!(boolean) db.getData(className + ALLOW_DOUBLEDOORS)) {
            vp.getLogger()
                    .warning("Double doors are disabled. ALLOW_DOUBLEDOORS value: "
                            + (boolean) db.getData(className + ALLOW_DOUBLEDOORS));
            return;
        }

        if (blockData instanceof Door) {
            Door door = this.getBottomDoor((Door) blockData, clickedBlock);
            Block otherDoorBlock = this.getOtherPart(door, clickedBlock);
            if (otherDoorBlock != null && otherDoorBlock.getBlockData() instanceof Door) {
                Door otherDoor = (Door) otherDoorBlock.getBlockData();
                this.toggleOtherDoor(clickedBlock, otherDoorBlock, !otherDoor.isOpen());
                autoClose.put(otherDoorBlock,
                        System.currentTimeMillis()
                                + Long.valueOf((long) db.getData(className + AUTOCLOSE_DELAY)) * 1000);
            }
        }
        autoClose.put(clickedBlock,
                System.currentTimeMillis() + Long.valueOf((long) db.getData(className + AUTOCLOSE_DELAY)) * 1000);
    }

    @EventHandler
    public void onKnock(PlayerInteractEvent e) {
        Player p = e.getPlayer();
        GameMode gm = p.getGameMode();

        if (gm == GameMode.CREATIVE || gm == GameMode.SPECTATOR)
            return;

        if (!p.hasPermission(PERMS.DOORSMODULE.KNOCK) || e.getAction() != Action.LEFT_CLICK_BLOCK
                || e.getHand() != EquipmentSlot.HAND)
            return;

        if ((boolean) db.getData(className + KNOCKING_REQUIRES_SHIFT) && !p.isSneaking())
            return;

        if ((boolean) db.getData(className + KNOCKING_REQUIRES_EMPTY_HAND)
                && p.getInventory().getItemInMainHand().getType() != Material.AIR)
            return;

        if (e.getClickedBlock() == null)
            return;

        Block block = e.getClickedBlock();
        BlockData blockData = block.getBlockData();

        if ((blockData instanceof Door && (boolean) db.getData(className + ALLOW_KNOCKING))
                || (blockData instanceof TrapDoor && (boolean) db.getData(className + ALLOW_KNOCKING_TRAPDOORS))
                || (blockData instanceof Gate && (boolean) db.getData(className + ALLOW_KNOCKING_GATES))) {
            this.playKnockSound(block);
        }
    }

    public void playKnockSound(Block block) {
        Location loc = block.getLocation();
        World world = block.getWorld();
        Sound sound = Optional
                .ofNullable(
                        Registry.SOUNDS
                                .get(NamespacedKey.minecraft((String) db.getData(className + SOUND_KNOCK_WOOD))))
                .orElse(Sound.ITEM_SHIELD_BLOCK);
        SoundCategory category = Enums
                .getIfPresent(SoundCategory.class,
                        (String) db.getData(className + SOUND_KNOCK_CATEGORY))
                .or(SoundCategory.BLOCKS);
        float volume = (int) db.getData(className + SOUND_KNOCK_VOLUME);
        float pitch = (int) db.getData(className + SOUND_KNOCK_PITCH);

        world.playSound(loc, sound, category, volume, pitch);
    }

    public static void toggleDoor(Block doorBlock, Openable openable, boolean open) {
        openable.setOpen(open);
        doorBlock.setBlockData(openable);
    }

    public Door getBottomDoor(Door door, Block block) {
        Block below = (door.getHalf() == Bisected.Half.BOTTOM) ? block : block.getRelative(BlockFace.DOWN);
        if (below.getType() == block.getType() && below.getBlockData() instanceof Door) {
            return (Door) below.getBlockData();
        }
        return null;
    }

    public Block getOtherPart(Door door, Block block) {
        if (door != null) {
            for (AdjacentBlockRecord neighbour : POSSIBLE_NEIGHBOURS) {
                Block relative = block.getRelative(neighbour.offsetX(), 0, neighbour.offsetZ());
                Door otherDoor = (relative.getBlockData() instanceof Door) ? (Door) relative.getBlockData() : null;
                if (otherDoor != null
                        && neighbour.facing() == door.getFacing()
                        && neighbour.hinge() == door.getHinge()
                        && relative.getType() == block.getType()
                        && otherDoor.getHinge() != neighbour.hinge()
                        && otherDoor.isOpen() == door.isOpen()
                        && otherDoor.getFacing() == neighbour.facing()) {
                    return relative;
                }
            }
        }
        return null;
    }

    public void toggleOtherDoor(Block block, Block otherBlock, boolean open) {
        if (!(block.getBlockData() instanceof Door) || !(otherBlock.getBlockData() instanceof Door))
            return;

        Door door = (Door) block.getBlockData();
        Door otherDoor = (Door) otherBlock.getBlockData();

        new BukkitRunnable() {
            @Override
            public void run() {
                if (!(otherBlock.getBlockData() instanceof Door))
                    return;
                Door newDoor = (Door) block.getBlockData();
                if (newDoor.isOpen() == door.isOpen()) {
                    return;
                }
                DoorsModule.toggleDoor(otherBlock, otherDoor, open);
            }
        }.runTaskLater(vp, 1L);
    }

    @Override
    public boolean isEnabled() {
        return (boolean) db.getData(className + ENABLE);
    }

    @Override
    public void config() {
        db.setData(className + ENABLE, true);
        db.setData(className + SOUND_KNOCK_CATEGORY, "BLOCKS");
        db.setData(className + SOUND_KNOCK_PITCH, 1.0);
        db.setData(className + SOUND_KNOCK_VOLUME, 1.0);
        db.setData(className + SOUND_KNOCK_WOOD, "entity_zombie_attack_wooden_door");
        db.setData(className + ALLOW_AUTOCLOSE, true);
        db.setData(className + ALLOW_DOUBLEDOORS, true);
        db.setData(className + ALLOW_KNOCKING, true);
        db.setData(className + ALLOW_KNOCKING_GATES, true);
        db.setData(className + ALLOW_KNOCKING_TRAPDOORS, true);
        db.setData(className + KNOCKING_REQUIRES_EMPTY_HAND, true);
        db.setData(className + KNOCKING_REQUIRES_SHIFT, false);
        db.setData(className + AUTOCLOSE_DELAY, 6);
    }
}

File: src/main/kotlin/org/xodium/vanillaplus/modules/DoorsModule.kt
package org.xodium.vanillaplus.modules

import net.kyori.adventure.sound.Sound
import org.bukkit.GameMode
import org.bukkit.Material
import org.bukkit.block.Block
import org.bukkit.block.BlockFace
import org.bukkit.block.data.Bisected
import org.bukkit.block.data.BlockData
import org.bukkit.block.data.Openable
import org.bukkit.block.data.type.Door
import org.bukkit.block.data.type.Gate
import org.bukkit.block.data.type.TrapDoor
import org.bukkit.entity.Player
import org.bukkit.event.Event
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.block.Action
import org.bukkit.event.player.PlayerInteractEvent
import org.bukkit.inventory.EquipmentSlot
import org.xodium.vanillaplus.VanillaPlus.Companion.instance
import org.xodium.vanillaplus.data.AdjacentBlockData
import org.xodium.vanillaplus.data.SoundData
import org.xodium.vanillaplus.interfaces.ModuleInterface
import org.bukkit.Sound as BukkitSound

/** Represents a module handling door mechanics within the system. */
class DoorsModule : ModuleInterface<DoorsModule.Config> {
    override val config: Config = Config()

    override fun enabled(): Boolean = config.enabled

    private val autoClose = mutableMapOf<Block, Long>()
    private val possibleNeighbours = listOf(
        AdjacentBlockData(0, -1, Door.Hinge.RIGHT, BlockFace.EAST),
        AdjacentBlockData(0, 1, Door.Hinge.LEFT, BlockFace.EAST),

        AdjacentBlockData(1, 0, Door.Hinge.RIGHT, BlockFace.SOUTH),
        AdjacentBlockData(-1, 0, Door.Hinge.LEFT, BlockFace.SOUTH),

        AdjacentBlockData(0, 1, Door.Hinge.RIGHT, BlockFace.WEST),
        AdjacentBlockData(0, -1, Door.Hinge.LEFT, BlockFace.WEST),

        AdjacentBlockData(-1, 0, Door.Hinge.RIGHT, BlockFace.NORTH),
        AdjacentBlockData(1, 0, Door.Hinge.LEFT, BlockFace.NORTH)
    )

    init {
        if (enabled()) {
            instance.server.scheduler.runTaskTimer(
                instance,
                Runnable {
                    autoClose.entries.removeIf { (block, time) ->
                        System.currentTimeMillis() >= time && handleAutoClose(block)
                    }
                },
                config.initDelay,
                config.interval
            )
        }
    }

    @EventHandler(priority = EventPriority.HIGH, ignoreCancelled = true)
    fun on(event: PlayerInteractEvent) {
        if (!enabled()) return

        val clickedBlock = event.clickedBlock ?: return
        if (!isValidInteraction(event)) return
        when (event.action) {
            Action.LEFT_CLICK_BLOCK -> handleLeftClick(event, clickedBlock)
            Action.RIGHT_CLICK_BLOCK -> handleRightClick(clickedBlock)
            else -> return
        }
    }

    /**
     * Toggles the state of the door or gate based on the provided block and openable data.
     * @param block The block representing the door or gate to be toggled.
     * @param openable The Openable data representing the door or gate.
     * @param open The desired state (open or closed) for the door or gate.
     */
    private fun toggleDoor(block: Block, openable: Openable, open: Boolean) {
        openable.isOpen = open
        block.blockData = openable
    }

    /**
     * Handles the auto-close functionality for doors and gates.
     * @param block The block representing the door or gate to be closed.
     * @return True if the block was successfully closed, false otherwise.
     */
    private fun handleAutoClose(block: Block): Boolean {
        val data = block.blockData as? Openable ?: return false
        if (!data.isOpen) return false
        when (data) {
            is Door -> handleDoorClose(block, data)
            is Gate -> handleGateClose(block)
            else -> return false
        }
        data.isOpen = false
        block.blockData = data
        return true
    }

    /**
     * Handles the sound effect for closing a door.
     * @param block The block representing the door being closed.
     * @param door The door block data.
     */
    private fun handleDoorClose(block: Block, door: Door) {
        getOtherPart(door, block)?.let { toggleOtherDoor(block, it, false) }
        block.world.playSound(config.soundDoorClose.toSound())
    }

    /**
     * Handles the sound effect for closing a gate.
     * @param block The block representing the gate being closed.
     */
    private fun handleGateClose(block: Block) =
        block.world.playSound(config.soundGateClose.toSound())

    /**
     * Checks if the interaction event is valid for processing.
     * @param event The `player interact event`.
     * @return True if the interaction is valid, false otherwise.
     */
    private fun isValidInteraction(event: PlayerInteractEvent): Boolean {
        return event.hand == EquipmentSlot.HAND &&
                event.useInteractedBlock() != Event.Result.DENY &&
                event.useItemInHand() != Event.Result.DENY
    }

    /**
     * Handles the left-click interaction with doors and gates, playing a knock sound if applicable.
     * @param event The `player interact event`.
     * @param block The block representing the door or gate being interacted with.
     */
    private fun handleLeftClick(event: PlayerInteractEvent, block: Block) {
        if (canKnock(event, event.player) && isKnockableBlock(block.blockData)) {
            block.world.playSound(config.soundKnock.toSound())
        }
    }

    /**
     * Handles the right-click interaction with doors and gates, toggling their state and handling auto-close functionality.
     * @param block The block representing the door or gate being interacted with.
     */
    private fun handleRightClick(block: Block) {
        if (config.allowDoubleDoors && (block.blockData is Door || block.blockData is Gate)) {
            processDoorOrGateInteraction(block)
        }
        if (config.allowAutoClose) autoClose[block] =
            System.currentTimeMillis() + config.autoCloseDelay
    }

    /**
     * Processes the interaction with doors or gates, toggling their state and handling auto-close functionality.
     * @param block The block representing the door or gate being interacted with.
     */
    private fun processDoorOrGateInteraction(block: Block) {
        val door = (block.blockData as? Door) ?: return
        val door2Block = getOtherPart(getDoorBottom(door, block), block) ?: return
        val secondDoor = door2Block.blockData as? Door ?: return
        toggleOtherDoor(block, door2Block, !secondDoor.isOpen)
        if (config.allowAutoClose) {
            autoClose[door2Block] = System.currentTimeMillis() + config.autoCloseDelay
        }
    }

    /**
     * Checks if the player can knock on the block based on their game mode and interaction conditions.
     * @param event The `player interact event`.
     * @param player The player attempting to knock.
     * @return True if the player can knock, false otherwise.
     */
    private fun canKnock(event: PlayerInteractEvent, player: Player): Boolean {
        return when {
            player.gameMode == GameMode.CREATIVE || player.gameMode == GameMode.SPECTATOR -> false
            event.action != Action.LEFT_CLICK_BLOCK || event.hand != EquipmentSlot.HAND -> false
            isKnockingConditionViolated(player) -> false
            else -> true
        }
    }

    /**
     * Checks if the knocking conditions are violated based on the player's state and configuration.
     * @param player The player attempting to knock.
     * @return True if the knocking conditions are violated, false otherwise.
     */
    private fun isKnockingConditionViolated(player: Player): Boolean {
        return (config.knockingRequiresShifting && !player.isSneaking) ||
                (config.knockingRequiresEmptyHand &&
                        player.inventory.itemInMainHand.type != Material.AIR)
    }

    /**
     * Checks if the block data is of a type that can be knocked on.
     * @param data The block data to check.
     * @return True if the block can be knocked on, false otherwise.
     */
    private fun isKnockableBlock(data: BlockData): Boolean {
        return when (data) {
            is Door -> config.allowKnockingDoors
            is Gate -> config.allowKnockingGates
            is TrapDoor -> config.allowKnockingTrapdoors
            else -> false
        }
    }

    /**
     * Toggles the state of the other door when one door is opened or closed.
     * @param block The block representing the first door.
     * @param block2 The block representing the second door.
     * @param open The desired state (open or closed) for the second door.
     */
    private fun toggleOtherDoor(block: Block, block2: Block, open: Boolean) {
        if (block.blockData !is Door || block2.blockData !is Door) return
        instance.server.scheduler.runTaskLater(
            instance,
            Runnable {
                val door = block.blockData as Door
                val door2 = block2.blockData as Door
                if (door.isOpen != door2.isOpen) toggleDoor(block2, door2, open)
            },
            config.initDelay
        )
    }

    /**
     * Retrieves the bottom half of a door if the provided block is the top half.
     * @param door The door block data.
     * @param block The block to check.
     * @return The bottom half of the door if it exists, null otherwise.
     */
    private fun getDoorBottom(door: Door, block: Block): Door? {
        return (if (door.half == Bisected.Half.BOTTOM) block else block.getRelative(BlockFace.DOWN))
            .blockData as? Door
    }

    /**
     * Retrieves the other part of a door if it exists.
     * @param door The door block data.
     * @param block The block to check.
     * @return The other part of the door if it exists, null otherwise.
     */
    private fun getOtherPart(door: Door?, block: Block): Block? {
        if (door == null) return null
        return possibleNeighbours.firstOrNull { neighbour ->
            val relative = block.getRelative(neighbour.offsetX, 0, neighbour.offsetZ)
            (relative.blockData as? Door)?.takeIf {
                it.facing == door.facing &&
                        it.hinge != door.hinge &&
                        it.isOpen == door.isOpen &&
                        relative.type == block.type
            } != null
        }?.let { block.getRelative(it.offsetX, 0, it.offsetZ) }
    }

    data class Config(
        override var enabled: Boolean = true,
        var initDelay: Long = 1L,
        var interval: Long = 1L,
        var allowAutoClose: Boolean = true,
        var allowDoubleDoors: Boolean = true,
        var allowKnockingDoors: Boolean = true,
        var allowKnockingGates: Boolean = true,
        var allowKnockingTrapdoors: Boolean = true,
        var knockingRequiresEmptyHand: Boolean = true,
        var knockingRequiresShifting: Boolean = true,
        var autoCloseDelay: Long = 6L * 1000L,
        var soundDoorClose: SoundData = SoundData(
            BukkitSound.BLOCK_IRON_DOOR_CLOSE,
            Sound.Source.BLOCK
        ),
        var soundGateClose: SoundData = SoundData(
            BukkitSound.BLOCK_FENCE_GATE_CLOSE,
            Sound.Source.BLOCK
        ),
        var soundKnock: SoundData = SoundData(
            BukkitSound.ENTITY_ZOMBIE_ATTACK_WOODEN_DOOR,
            Sound.Source.HOSTILE
        ),
    ) : ModuleInterface.Config
}
